lib/tests/it/main.rs

```diff
@@ -601,6 +601,99 @@ async fn impl_test_container_import_export(chunked: bool) -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn test_export_as_container_nonderived() -> Result<()> {
+    let fixture = Fixture::new_v1()?;
+    // Export into an OCI directory
+    let src_imgref = fixture.export_container().await.unwrap().0;
+
+    let initimport = fixture.must_import(&src_imgref).await?;
+    let initimport_ls = fixture::ostree_ls(fixture.destrepo(), &initimport.merge_commit).unwrap();
+
+    let exported_ocidir_name = "exported.ocidir";
+    let dest = ImageReference {
+        transport: Transport::OciDir,
+        name: format!("{}:exported-test", fixture.path.join(exported_ocidir_name)),
+    };
+    fixture.dir.create_dir(exported_ocidir_name)?;
+    let ocidir = ocidir::OciDir::ensure(&fixture.dir.open_dir(exported_ocidir_name)?)?;
+    let exported = store::export(fixture.destrepo(), &src_imgref, &dest, None)
+        .await
+        .unwrap();
+    let (new_manifest, desc) = ocidir.read_manifest_and_descriptor()?;
+    assert_eq!(desc.digest(), exported.as_str());
+    assert_eq!(new_manifest.layers().len(), fixture::LAYERS_V0_LEN);
+
+    // Reset the destrepo
+    fixture.clear_destrepo()?;
+    // Clear out the original source
+    std::fs::remove_dir_all(src_imgref.name.as_str())?;
+
+    let reimported = fixture.must_import(&dest).await?;
+    let reimport_ls = fixture::ostree_ls(fixture.destrepo(), &reimported.merge_commit).unwrap();
+    similar_asserts::assert_eq!(initimport_ls, reimport_ls);
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_export_as_container_derived() -> Result<()> {
+    let fixture = Fixture::new_v1()?;
+    // Export into an OCI directory
+    let src_imgref = fixture.export_container().await.unwrap().0;
+    // Add a derived layer
+    let derived_tag = "derived";
+    // Build a derived image
+    let srcpath = src_imgref.name.as_str();
+    let temproot = &fixture.path.join("temproot");
+    || -> Result<_> {
```

Comment:
This looks like magic to me... the examples I see of closures assign to a variable. Is this using a closure to use the variables you just declared before instead of writing a function and passing variables to it?

Reply:
BTW, totally fine asking questions like this on PRs, please continue to do so!

This is an [IIFE](https://en.wikipedia.org/wiki/Immediately_invoked_function_expression) - it's...somewhat unusual. The rationale here is actually so that `    .context("generating temp content")?;` applies to *everything* inside the closure - it's a way of adding context to a whole bunch of `?` without actually splitting it out into a separate function as you say.

Splitting into a separate function is more heavyweight: requires spelling out all the types and passing all the inputs.

But yes, I don't see IIFEs in much other Rust code, and I don't tend to use them often.  If I remember right, I added this one because I had a bug *somewhere* in all those lines and I didn't want to change everything to use `unwrap()`.

Reply:
thanks that makes sense.

Reply:
That said, I actually copy-pasted this, so https://github.com/ostreedev/ostree-rs-ext/pull/643 dedups it